pkg/apis/storage/types.go

```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
 	// Items is the list of StorageClasses
 	Items []StorageClass
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Captures the intent to attach or detach the specified volume to/from
+// the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+	metav1.TypeMeta
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the desired attach/detach volume behavior.
+	// Populated by the Kubernetes system.
+	Spec VolumeAttachmentSpec
+
+	// Status of the VolumeAttachment request.
+	// Populated by the entity completing the attach or detach
+	// operation, i.e. the external-attacher.
+	Status VolumeAttachmentStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeAttachmentList is a collection of VolumeAttachment objects.
+type VolumeAttachmentList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is the list of VolumeAttachments
+	Items []VolumeAttachment
+}
+
+// The specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+	// The name of the external-attacher that MUST handle
+	// this request.
+	Attacher string
+
+	// The volume to attach.
+	Volume api.VolumeSource
+
+	// The node that the volume should be attached to.
+	NodeName string
```

Comment:
How is the VolumeAttachment object used during a multi-attachments scenario? For example with RO volumes, will there be multiple of these VolumeAttachment objects for every RO "attachment" even if the volume source and node is the same or will there just be one?

Reply:
Should we also store information about pods and mountDir in the case where plugins are not using the attacher-detacher controller?

Reply:
> volumes, will there be multiple of these VolumeAttachment objects for every RO "attachment" even if the volume source and node is the same or will there just be one?

We do not plan to change current behavior, i.e. a volume is attached to a node only once. So there will be one instance per node-volume pair. Kubelet then mounts the single device to individual pods (using bind-mounts).

> Should you also store information about pods and mountDir in the case where plugins are not using the attacher-detacher controller?

I am not sure I understand the question. That's kubelet job and is not related to volume attachment.

Reply:
I was alluding to the fact that in non-attachable plugins, kubelet mount() does everything (attach, mountDevice and mount podDir). So my question was if there is a need to store pod information to determine when it is safe to detach (delete VolumeAttach object).

Look like kubelet can determine how many bind mount ref are there during a detach and delete the attachment once there are no bind mount left.

Reply:
`VolumeAttachment` is used only for attachable plugins (CSI). See the [proposal](https://github.com/kubernetes/community/pull/1258) for handling of CSI plugins that don't support `ControllerPublishVolume` call.

And indeed, kubelet has its own means how to track pods that use the volume and it won't let Attach/Detach controller to detach a volume that's still mounted.